Dynamically check for supported FourCC on Windows.

mmuetzel opened the pull request:
Instead of hard-coding a list of `MFVideoFormat_` constant, generate a GUID from the FourCC on runtime. If it isn't actually supported by the device, it will fail later.

With it, I see the following on Windows 11 with a webcam:
```
                                                                                                                                                                  
>> set(obj,"VideoFormat").fourcc

ans = H264
ans = MJPG
ans = NV12
ans = YUY2
>> k = 1

k = 1
>> fmts = {set(obj,"VideoFormat").fourcc};

>> set (obj, "VideoFormat", fmts{k});

>> set (obj, "VideoFormat", 'abcd');
cl_mf_handler.cc:s_fmt:458 failed with 0
error: mf_handler::s_fmt (abcd, 0, 0) SetCurrentMediaType failed
error: called from
    set at line 101 column 15
>> set (obj, "VideoFormat", 'abcde');
error: FourCC must have exactly 4 characters.
error: called from
    set at line 101 column 15
>>
```

mmuetzel committed:
Dynamically check for supported FourCC on Windows.

Andy1978 reviewed (approved):
Looks great, I haven't found a way to map fourCC to MFVideoFormat_ yet.

Perhaps a ::toupper for fourCC would also be a good idea and there are some
values, which can't be directly mapped like

"HEVS" ->  MFVideoFormat_HEVC_ES

mmuetzel commented:
Yeah. I realized a bit too late that I reverted some mappings when I rebased on your latest change.

I'm not sure about the ::toupper of the FourCC. There are a couple of GUIDs that have the FourCC in lower case:
https://github.com/mingw-w64/mingw-w64/blob/c67e9a7ab6d4551c65f7d81dccbd8826e4b5ad78/mingw-w64-headers/include/mfapi.h#L399-L405
```
  DEFINE_MEDIATYPE_GUID (MFVideoFormat_DVSL, FCC ('dvsl'));
  DEFINE_MEDIATYPE_GUID (MFVideoFormat_DVSD, FCC ('dvsd'));
  DEFINE_MEDIATYPE_GUID (MFVideoFormat_DVHD, FCC ('dvhd'));
  DEFINE_MEDIATYPE_GUID (MFVideoFormat_DV25, FCC ('dv25'));
  DEFINE_MEDIATYPE_GUID (MFVideoFormat_DV50, FCC ('dv50'));
  DEFINE_MEDIATYPE_GUID (MFVideoFormat_DVH1, FCC ('dvh1'));
  DEFINE_MEDIATYPE_GUID (MFVideoFormat_DVC, FCC ('dvc '));
```

Unfortunately, the function that re-implements these macros on run-time is case-sensitive. I'm not sure how to best deal with that...

mmuetzel commented:
There's also these for which the suffix is lower case:
```
  DEFINE_MEDIATYPE_GUID (MFVideoFormat_v210, FCC ('v210'));
  DEFINE_MEDIATYPE_GUID (MFVideoFormat_v216, FCC ('v216'));
  DEFINE_MEDIATYPE_GUID (MFVideoFormat_v410, FCC ('v410'));
```

Andy1978 commented:
One thing to consider is also, that it's now possible to set many new formats (like H264 which your webcam supports) but there is no converter to tranform them to RGB yet. What happens if you execute
```
set (obj, "VideoFormat", "H264");
img = getsnapshot (obj);
```

I only have a limited  set on webcams, no one which supports H264. Btw, what is your model?

mmuetzel commented:
Good point. That fails:
```
>> pkg load image-acquisition
>> obj = videoinput (__test__device__{:});
>> set (obj, "VideoFormat", "H264");
>> img = getsnapshot (obj);
error: mf_handler::capture: no conversion from 'H264' to RGB3 implemented yet
error: called from
    getsnapshot at line 78 column 6
>>
```

I have a Logitech C920.

mmuetzel commented:
Fwiw, the number of test failures reduces to 1 for me after this change.
```
Failure Summary:

  ..60\packages\image-acquisition-0.3.2\@videoinput\getsnapshot.m  pass    2/3
                                                                   FAIL    1

Summary:

  PASS                               17
```

The error is exactly what you predicted:
```
>>>>> processing C:\Users\Markus\AppData\Roaming\octave\api-v60\packages\image-acquisition-0.3.2\@videoinput\getsnapshot.m
***** test
 obj = videoinput (__test__device__{:});
 fmts = {set(obj,"VideoFormat").fourcc};
 for k = 1:numel (fmts)
   set (obj, "VideoFormat", fmts{k});
   s = get (obj, "VideoResolution");
   start (obj)
   img = getsnapshot (obj);
   assert (size (img), [fliplr(s) 3]);
   stop (obj)
 endfor
!!!!! test failed
mf_handler::capture: no conversion from 'H264' to RGB3 implemented yet
```

Andy1978 commented:
I'm unsure what to do here. Historically I only implemented v4l2 where v4l2 itself can convert between many formats. Then I tried to also implement Media Foundation (because Direct Show is called deprecated).

My thought was to give the user the possibility to capture the stream as raw as possible (YUYV, YUV422, YV12 for example with different resolution for Y and Cb+Cr) but also offer the opportunity to automatically convert to RGB3 for further processing with the image package.

I think I should change the test as follows:
- get a list with possible formats supported by the driver
- set each of them
- try to get a snapshot and convert to RGB3, if this fails (catch "no conversion") get a raw representation

mmuetzel commented:
> I think I should change the test as follows:
> 
> * get a list with possible formats supported by the driver
> * set each of them
> * try to get a snapshot and convert to RGB3, if this fails (catch "no conversion") get a raw representation

I think that is a reasonable approach. Sounds good to me. 👍

Andy1978 commented:
@mmuetzel Can you try the lastest patch? I was able to get a Logitech C922 Pro Stream in the meanwhile but this doesn't support H264 so I'm not able to test this...

mmuetzel commented:
Thanks for looking into this.

With the latest head (8e04baa89adcc19920182663fe22b4627776e89d), the tests pass with my webcam on Windows 11 using Octave 10.1.0:
```
Summary:

  PASS                               17
  FAIL                                0
```

I see the following when trying to use an unsupported encoding:
```
>> obj = videoinput (__test__device__{:});
>> set (obj, "VideoFormat").fourcc

ans = H264
ans = MJPG
ans = NV12
ans = YUY2
>> k = 1

k = 1
>> fmts = {set(obj, "VideoFormat").fourcc};
>> set (obj, "VideoFormat", fmts{k});
>> img = getsnapshot (obj);

error: mf_handler::capture: no conversion from 'H264' to RGB3 implemented yet, please try another VideoFormat
 or use RAW = true.
error: called from
    getsnapshot at line 83 column 6
```

That error message looks informative to me. 👍

Are you planning to release a tarball containing these changes? I don't know yet when Octave 10.2.0 is going to be released. (Probably, not too long from now.) Maybe, we could include your package in MXE Octave before that...

mmuetzel commented:
Aah. Forgot to show.
A supported encoding works fine.
In the same session as above:
```
>> k = 4
k = 4
>> set (obj, "VideoFormat", fmts{k});
>> img = getsnapshot (obj);
>> image (img);
```
And the displayed image looks correct.

Andy1978 commented:
> Are you planning to release a tarball containing these changes? I don't know yet when Octave 10.2.0 is going to be released. (Probably, not too long from now.) Maybe, we could include your package in MXE Octave before that...

The latest 0.3.3 release includes the changes.
